lib/eigensolve_quda.cpp

```diff
@@ -409,18 +462,58 @@ namespace quda
 #ifdef HAVE_QIO
     const int Nvec = eig_vecs.size();
     std::vector<ColorSpinorField *> tmp;
+    tmp.reserve(Nvec);
+    auto spinor_parity = eig_vecs[0]->SuggestedParity();
     if (eig_vecs[0]->Location() == QUDA_CUDA_FIELD_LOCATION) {
       ColorSpinorParam csParam(*eig_vecs[0]);
       csParam.fieldOrder = QUDA_SPACE_SPIN_COLOR_FIELD_ORDER;
       csParam.setPrecision(eig_vecs[0]->Precision() < QUDA_SINGLE_PRECISION ? QUDA_SINGLE_PRECISION :
                                                                               eig_vecs[0]->Precision());
       csParam.location = QUDA_CPU_FIELD_LOCATION;
-      for (int i = 0; i < Nvec; i++) {
-        tmp.push_back(ColorSpinorField::Create(csParam));
-        *tmp[i] = *eig_vecs[i];
+
+      if (csParam.siteSubset == QUDA_FULL_SITE_SUBSET || csParam.nColor != 3) { // we don't care for MG vectors
+        // We're good, copy as is.
+        csParam.create = QUDA_NULL_FIELD_CREATE;
+        for (int i = 0; i < Nvec; i++) {
+          tmp.push_back(ColorSpinorField::Create(csParam));
+          *tmp[i] = *eig_vecs[i];
+        }
+      } else { // QUDA_PARITY_SITE_SUBSET
+        csParam.create = QUDA_NULL_FIELD_CREATE;
+
+        // intermediate host single-parity field
+        ColorSpinorField *tmp_intermediate = ColorSpinorField::Create(csParam);
+
+        csParam.x[0] *= 2;                          // corrects for the factor of two in the X direction
+        csParam.siteSubset = QUDA_FULL_SITE_SUBSET; // create a full-parity field.
+        csParam.create = QUDA_ZERO_FIELD_CREATE;    // to explicitly zero the odd sites.
+        for (int i = 0; i < Nvec; i++) {
+          tmp.push_back(ColorSpinorField::Create(csParam));
+
+          // copy the single parity eigen/singular vector into an
+          // intermediate device-side vector
+          *tmp_intermediate = *eig_vecs[i];
+
+          // copy the single parity only eigen/singular vector into the even components of the full parity vector
+          if (spinor_parity == QUDA_EVEN_PARITY)
+            blas::copy(tmp[i]->Even(), *tmp_intermediate);
+          else if (spinor_parity == QUDA_ODD_PARITY)
+            blas::copy(tmp[i]->Odd(), *tmp_intermediate);
+          else
+            errorQuda("When saving single parity vectors, the suggested parity must be set.");
+        }
+        delete tmp_intermediate;
       }
     } else {
-      for (int i = 0; i < Nvec; i++) { tmp.push_back(eig_vecs[i]); }
+      // FIXME: add conversion to full site subset from single parity fields.
```

Comment:
what does this FIXME refer to?

Reply:
Outdated, my original hack didn't support fixing single-parity CPU fields. Removed.

Reply:
Nevermind, I never actually finished fixing that! Good catch. Addressing it now.